src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs

```diff
@@ -68,6 +69,37 @@ internal WebAssemblyHostBuilder(WebAssemblyJSRuntimeInvoker jsRuntimeInvoker)
             };
         }
 
+        private void InitializeRegisteredRootComponents(WebAssemblyJSRuntimeInvoker jsRuntimeInvoker)
+        {
+            var componentsCount = jsRuntimeInvoker.InvokeUnmarshalled<object, object, object, int>(RegisteredComponentsInterop.GetRegisteredComponentsCount, null, null, null);
+            if (componentsCount == 0)
+            {
+                return;
+            }
+
+            var registeredComponents = new ClientComponentMarker[componentsCount];
+            for (var i = 0; i < componentsCount; i++)
+            {
+                var id = jsRuntimeInvoker.InvokeUnmarshalled<int, object, object, int>(RegisteredComponentsInterop.GetId, i, null, null);
```

Comment:
Is it possible to read these in one call?

Reply:
AFAIK I believe we can't. I know there are functions to read from the heap, but I don't think there are functions to write to the heap. Are there?

I don't think this is incredibly critical since unmarshalled interop should be pretty fast, I'm happy to change this post RC1 if we think it's important.

Reply:
Personally I don't feel strongly about this, as it's reasonably clear what's going on from the code.

If we wanted to simplify it on the .NET side, we might consider having a single *marshalled* call that returns the whole set of descriptors. The extra perf impact of that would be marginal because we're doing JSON deserialization for the parameters anyway. Marshalled calls are only slower because of the JSON aspect - in other respects they are basically the same as unmarshalled calls.

Reply:
We'd have to measure it to be certain, but it *might* actually be faster as a single marshalled call here because we'd only have to go through the process of copying string data from JS to .NET once, instead of 5x for each root component.

However I still don't think it's urgent to optimize this.